MetaMask/metamask-extension
# e2e: better split of tests running on parallel machines

## **TL;DR**

It makes the E2E workflow faster by a better split of tests running on parallel machines.

**Major long-standing consequence:** you will see individual tests failing whilst the workflow still passes.  The CI is set to `--retries 2`, so it might fail 1 of the 3 runs of an individual test, but still pass the job.  (Before this PR, it wouldn't even report that it failed and retried, it would just say `Success`)  This means that more truly "Flaky" tests will be surfaced.

## **Details**

This PR continues the work of #16417 and #16455

It follows the procedures from both of these CircleCI docs pages to split tests by timing data, and also to enable that "Rerun failed tests" button:
- https://circleci.com/docs/use-the-circleci-cli-to-split-tests
- https://circleci.com/docs/rerun-failed-tests

### Package changes

The XML reporter we were using before, `xunit`, was not reporting the filename of the spec test, so I added the devDepencency `mocha-junit-reporter`.  This also allowed me to eliminate `junit-report-merger`, the `Merge JUnit report` steps in `config.yml`, and the file `e2e-process-report.js`.

### run-all.js

This is the file with the most major changes.  It takes the next step in what @FrederikBolding did for #16417, but instead of dividing the parallelism equally by test file, it stores information about how long each test took recently, and divides by timing.  The only way I could figure out how to pass info between command line steps was to write to and read from files, so I made `test/test-results/fullTestList.txt` and `test/test-results/myTestList.txt`.

The console will show you how the test files are getting split, with output like this:
```
Full test list: /home/circleci/project/test/e2e/snaps/test-snap-bip-32.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-bip-44.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-cronjob.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-dialog.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-ethprovider.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-get-locale.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-getentropy.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-installed.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-lifecycle.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-management.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-managestate.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-networkaccess.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-notification.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-revoke-perm.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-rpc.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-txinsights.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-update.spec.js
/home/circleci/project/test/e2e/snaps/test-snap-wasm.spec.js
/home/circleci/project/test/e2e/accounts/test-create-snap-account.spec.js
/home/circleci/project/test/e2e/accounts/test-remove-accounts-snap.spec.js
/home/circleci/project/test/e2e/accounts/test-snap-accounts.spec.js
/home/circleci/project/test/e2e/flask/petnames.spec.js
/home/circleci/project/test/e2e/flask/ppom-blockaid-alert-erc20-approval.spec.js
/home/circleci/project/test/e2e/flask/ppom-blockaid-alert-erc20-transfer.spec.js
/home/circleci/project/test/e2e/flask/ppom-blockaid-alert.spec.js
/home/circleci/project/test/e2e/flask/ppom-toggle-settings.spec.js
My test list: [
  '/home/circleci/project/test/e2e/snaps/test-snap-cronjob.spec.js',
  '/home/circleci/project/test/e2e/accounts/test-create-snap-account.spec.js',
  '/home/circleci/project/test/e2e/snaps/test-snap-dialog.spec.js',
  '/home/circleci/project/test/e2e/snaps/test-snap-getentropy.spec.js',
  '/home/circleci/project/test/e2e/snaps/test-snap-txinsights.spec.js',
  '/home/circleci/project/test/e2e/snaps/test-snap-revoke-perm.spec.js',
  '/home/circleci/project/test/e2e/snaps/test-snap-rpc.spec.js',
  '/home/circleci/project/test/e2e/snaps/test-snap-installed.spec.js',
  '/home/circleci/project/test/e2e/flask/ppom-blockaid-alert-erc20-approval.spec.js',
  '/home/circleci/project/test/e2e/snaps/test-snap-networkaccess.spec.js\n'
]
```

### Splitting inside one test file
_**Q:**_ Can you theoretically split tests inside a single test file?
_**A:**_ Theoretically, yes, but it would require more extensive changes to how we run tests.

### To be continued

I myself am a huge offender in this CI runtime creep with `test-snap-accounts.spec.js`, which can take up to 36 minutes to run all by itself.  Independent from this PR, I am going to make changes to reduce that runtime, but these two will be symbiotic.

## **Manual testing steps**

1. Look at a [Workflow on this PR](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/60196/workflows/3ddf7bc6-6eea-478d-a2a0-a79ce8b8383d) to see how it's splitting and reporting tests differently now
2. Look at the difference in timing from develop, and how it's bringing the total time down
3. Find a Job where a test failed but the Job still passed
4. Look for the "Rerun failed tests" button and the `FAILED TESTS RERUN` on [the CircleCI history of this branch](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension?branch=e2e%2Ftest-splitting) 
5. Find a `FAILED TESTS RERUN` that shows `run-all.js info: Skipping this node because there is no myTestList.txt` on some of the nodes

## **Screenshots/Recordings**

### **Before**
### ${\color{red}Parallel \space run \space timing \space before}$
![image](https://github.com/MetaMask/metamask-extension/assets/539738/4a50639f-4aa8-4738-9ca3-95b6b8c1f8fe)

### **After**

### ${\color{red}Parallel \space run \space timing \space after}$
![image](https://github.com/MetaMask/metamask-extension/assets/539738/a04f8af3-f9fb-4f15-babd-076ba72e3036)

### ${\color{red}Example \space of \space Success \space with \space failed \space tests}$
![image](https://github.com/MetaMask/metamask-extension/assets/539738/9121ce23-9c2e-41a3-92b5-c6b21d7381d6)

### ${\color{red}FAILED \space TESTS \space RERUN}$
![image](https://github.com/MetaMask/metamask-extension/assets/539738/b6bfdbdc-2d77-4312-b959-839a36fd2547)

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
